### PR TITLE
Move maps to join partition

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: ref #6233

Problem Summary:

### What is changed and how it works?
1. move hash map to `JoinPartition`
2. move all the insert/probe hash table code to `JoinPartition`
3. refine `NonJoinedBlockInputStream` so it now read non joined data based on `JoinPartition`

blocked by #7127
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/DataStreams/NonJoinedBlockInputStream.cpp [hunk 1]
```diff
@@ -237,66 +270,32 @@ size_t NonJoinedBlockInputStream::fillColumns(const Map & map,
...
         if ((*it)->getMapped().getUsed())
         {
             ++(*it);
             continue;
         }
 
-        rows_added += AdderNonJoined<STRICTNESS, typename Map::mapped_type>::add((*it)->getMapped(), key_num, num_columns_left, mutable_columns_left, num_columns_right, mutable_columns_right, next_element_in_row_list, max_block_size - rows_added);
-        assert(rows_added <= max_block_size);
+        rows_added += AdderNonJoined<STRICTNESS, typename Map::mapped_type>::add((*it)->getMapped(), key_num, num_columns_left, mutable_columns_left, num_columns_right, mutable_columns_right, next_element_in_row_list, max_block_size - rows_added - rows_before_fill);
+        assert(rows_added + rows_before_fill <= max_block_size);
```
Comment (yibin87) [inline dbms/src/DataStreams/NonJoinedBlockInputStream.cpp:298]: move assert before AdderNonJoined::add statement? And can add a new local variable to record "rows_added + rows_before_fill"

**Review by yibin87 — APPROVED** [review]
Others LGTM

File: dbms/src/Interpreters/JoinUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Interpreters/JoinUtils.h>
+
+#include "Flash/Mpp/HashBaseWriterHelper.h"
```
Comment (SeaRise) [inline dbms/src/Interpreters/JoinUtils.cpp:17]:
```suggestion
#include <Flash/Mpp/HashBaseWriterHelper.h>
```

File: dbms/src/DataStreams/NonJoinedBlockInputStream.cpp (hunk 1 above)
Comment (windtalker) [reply dbms/src/DataStreams/NonJoinedBlockInputStream.cpp]: Done

File: dbms/src/Interpreters/JoinUtils.cpp (hunk 2 above)
Comment (windtalker) [reply dbms/src/Interpreters/JoinUtils.cpp]: Done

**Review by SeaRise — APPROVED** [review]

File: dbms/src/Interpreters/Join.h [hunk 3]
```diff
@@ -485,29 +340,36 @@ class Join
...
+    void joinBlockCrossImpl(Block & block, ConstNullMapPtr null_map) const;
+
+    template <ASTTableJoin::Kind KIND, ASTTableJoin::Strictness STRICTNESS, typename Maps>
+    void joinBlockNullAwareImpl(
+        Block & block,
+        size_t left_columns,
+        const ColumnRawPtrs & key_columns,
+        const ConstNullMapPtr & null_map,
+        const ConstNullMapPtr & filter_map,
+        const ConstNullMapPtr all_key_null_map) const;
```
Comment (gengliqi) [inline dbms/src/Interpreters/Join.h:365]:
```suggestion
        const ConstNullMapPtr & all_key_null_map) const;
```

File: dbms/src/Interpreters/JoinHashTable.h [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    using key_strbinType = HashMapWithSavedHash<StringRef, Mapped>;
+    std::unique_ptr<key_strbinType> key_strbin;
+    using key_fixed_stringType = HashMapWithSavedHash<StringRef, Mapped>;
+    std::unique_ptr<key_fixed_stringType> key_fixed_string;
+    using keys128Type = HashMap<UInt128, Mapped, HashCRC32<UInt128>>;
+    std::unique_ptr<keys128Type> keys128;
+    using keys256Type = HashMap<UInt256, Mapped, HashCRC32<UInt256>>;
+    std::unique_ptr<keys256Type> keys256;
+    using serializedType = HashMapWithSavedHash<StringRef, Mapped>;
+    std::unique_ptr<serializedType> serialized;
```
Comment (gengliqi) [inline dbms/src/Interpreters/JoinHashTable.h:141]: I think it's even less readable than before.
If we want to use type definitions, how about separating all type definitions from all `unique_ptr`?

File: dbms/src/Interpreters/Join.cpp [hunk 5]
```diff
@@ -2285,198 +1305,157 @@ void Join::joinBlockImplCross(Block & block) const
...
     const ColumnRawPtrs & key_columns,
-    const Sizes & key_sizes,
     const ConstNullMapPtr & null_map,
     const ConstNullMapPtr & filter_map,
-    const ConstNullMapPtr & all_key_null_map,
-    const TiDB::TiDBCollators & collators,
-    bool null_key_check_all_blocks_directly,
-    bool right_has_all_key_null_row,
-    bool right_table_is_empty)
+    const ConstNullMapPtr all_key_null_map) const
```
Comment (gengliqi) [inline dbms/src/Interpreters/Join.cpp:1430]:
```suggestion
    const ConstNullMapPtr & all_key_null_map) const
```

File: dbms/src/Interpreters/Join.cpp [hunk 6]
```diff
@@ -2877,73 +1711,17 @@ Block Join::joinBlock(ProbeProcessInfo & probe_process_info, bool dry_run) const
...
-    else if (kind == NullAware_LeftSemi && strictness == Any)
-        joinBlockImplNullAware<NullAware_LeftSemi, Any>(block, maps_any);
-    else if (kind == NullAware_LeftAnti && strictness == All)
-        joinBlockImplNullAware<NullAware_LeftAnti, All>(block, maps_all);
-    else if (kind == NullAware_LeftAnti && strictness == Any)
-        joinBlockImplNullAware<NullAware_LeftAnti, Any>(block, maps_any);
+    if (isCrossJoin(kind))
+        block = joinBlockCross(probe_process_info);
+    else if (!isNullAwareSemiFamily(kind))
+        block = joinBlockHash(probe_process_info);
```
Comment (gengliqi) [inline dbms/src/Interpreters/Join.cpp:1722]: How about 
```suggestion
    else if (isNullAwareSemiFamily(kind))
        block = joinBlockNullAware(probe_process_info);
    else 
        block = joinBlockHash(probe_process_info);
```

File: dbms/src/Interpreters/JoinHashTable.h (hunk 4 above)
Comment (windtalker) [reply dbms/src/Interpreters/JoinHashTable.h]: ok

File: dbms/src/Interpreters/JoinHashTable.h [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    std::unique_ptr<ConcurrentHashMapWithSavedHash<StringRef, Mapped>> key_strbin;
+    std::unique_ptr<ConcurrentHashMapWithSavedHash<StringRef, Mapped>> key_fixed_string;
+    std::unique_ptr<ConcurrentHashMap<UInt128, Mapped, HashCRC32<UInt128>>> keys128;
+    std::unique_ptr<ConcurrentHashMap<UInt256, Mapped, HashCRC32<UInt256>>> keys256;
+    std::unique_ptr<ConcurrentHashMapWithSavedHash<StringRef, Mapped>> serialized;
+    // TODO: add more cases like Aggregator
+};
+
+template <typename Mapped>
+struct MapsTemplate
```
Comment (gengliqi) [inline dbms/src/Interpreters/JoinHashTable.h:117]: The code style between `MapsTemplate` and `ConcurrentMapsTemplate` is different. I am ok with these two styles but they should be the same.

**Review by gengliqi — APPROVED** [review]
Rest LGTM

File: dbms/src/Interpreters/JoinHashTable.h (hunk 7 above)
Comment (windtalker) [reply dbms/src/Interpreters/JoinHashTable.h:130]: Updated.

**Pull request merged**